### PR TITLE
API change for zmq3 (4.* version on repo; tested on Ubuntu 16)

### PR DESCRIPTION
**Pull request by ustiugov**

### CODE REVIEW

**Comment by prekageo** [comment]
Hi Dmitrii,

Can you let me know which problem you are trying to solve with this pull request? I am able to compile the source code as it is after installing the libzmq3-dev package on Ubuntu. It doesn't work for you?

**Comment by ustiugov** [comment]
Hi George,
The build is broken on Ubuntu 14 and 16 due to libzmq package version problems. For example, here is the compilation error on Ubuntu 16 with libzmq3 installed (tested on icnals10 and one of the iccluster nodes):
```
g++ -o common.o -c -std=c++0x -D_GNU_SOURCE -O3 -Wall -g -I/usr/local/include -I/opt/local/include common.cc                                                                                               
common.cc: In function ‘void connect_agent()’:                                                                                                            
common.cc:349:38: error: no matching function for call to ‘zmq::socket_t::setsockopt(int, int)’                                                                                                            
     s->setsockopt(ZMQ_RCVTIMEO, 10000);                                                                                                                                                                   
                                      ^                                                                                                                                                                    
In file included from common.h:3:0,                                                                                                                                             
                 from common.cc:3:                                                                                                                                               
/usr/include/zmq.hpp:354:21: note: candidate: void zmq::socket_t::setsockopt(int, const void*, size_t)                                                                  
         inline void setsockopt (int option_, const void *optval_,                                                                                                     
                     ^~~~~~~~~~                                                                                                                                        
/usr/include/zmq.hpp:354:21: note:   candidate expects 3 arguments, 2 provided        
```
and after fixing it comes this one:
```
g++ -o mutilateudp.o -c -std=c++0x -D_GNU_SOURCE -O3 -Wall -g -I/usr/local/include -I/opt/local/include mutilateudp.cc
mutilateudp.cc: In function ‘void agent()’:
mutilateudp.cc:140:54: error: no matching function for call to ‘zmq::socket_t::bind(std::basic_string<char>)’
  socket.bind(string("tcp://*:") + args.agent_port_arg);
                                                      ^
In file included from mutilateudp.cc:4:0:
/usr/include/zmq.hpp:370:21: note: candidate: void zmq::socket_t::bind(const char*)
         inline void bind (const char *addr_)
                     ^~~~
/usr/include/zmq.hpp:370:21: note:   no known conversion for argument 1 from ‘std::basic_string<char>’ to ‘const char*’
```

Marios reproduced with a *newly* cloned repo the first error on icnals10. My guess is that the system got upgraded so the package `libzmq-dev` got a new version on that platform.  I had to fix a couple of incompatible API calls so that mutilate works with `libzmq3` package on Ubuntu 14.

**Comment by prekageo** [comment]
I see what might be the problem. Check if you have installed the libzmq-dev package:

```dpkg --get-selections | grep libzmq```

If yes, I suggest that you remove all libzmq packages and install only libzmq3-dev.

I agree that the situation with the zmq dependency is confusing. I have been confused myself many times!

**Comment by ustiugov** [comment]
George,
I have the following output on my Ubuntu 14 node:
```ustiugov@iccluster041:~/faas_study/mutilate$ dpkg --get-selections | grep libzmq                                          
libzmq1:amd64                                   deinstall                                                                 
libzmq3:amd64                                   install                                                       
libzmq3-dev:amd64                               install  
```
The compilation error is still there even though other libs are removed. Are you sure you tried a clean build?

**Comment by prekageo** [comment]
You are right. Your patch is necessary on Ubuntu Trusty. I have some long standing pending patches that I'd like to push first. I'll ask you to update your pull request once I do the push. Thanks!

**Comment by ustiugov** [comment]
Please go ahead, no rush. Thanks!

On 13 Mar 2019, at 14:35, George Prekas <notifications@github.com<mailto:notifications@github.com>> wrote:


You are right. Your patch is necessary on Ubuntu Trusty. I have some long standing pending patches that I'd like to push first. I'll ask you to update your pull request once I do the push. Thanks!

File: common.cc [hunk 1]
```diff
@@ -346,7 +346,9 @@ void connect_agent(void) {
     string host = string("tcp://") + string(args.agent_arg[i]) +
       string(":") + string(args.agent_port_arg);
     s->connect(host.c_str());
-    s->setsockopt(ZMQ_RCVTIMEO, 10000);
+    int64_t timeout = 10000;
```
Comment (prekageo) [inline common.cc]:
```suggestion
    int timeout = 10000;
```

The parameter must be int. It is checked by zmq.

**Review by prekageo — CHANGES_REQUESTED** [review]
Please rebase on top of the new master. Squash your commits to a single one with a message like "fix zmq for older Ubuntu".

Thanks for your effort!

**Comment by ustiugov** [comment]
ok, ready for the merge. Thanks!

**Pull request merged**